pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -0,0 +1,367 @@
+			if err != nil {
+				return nil
+			}
+			account = wallet.Accounts[len(wallet.Accounts)-1]
+		}
+	}
+
+	return &Service{
+		chain:             chain,
+		log:               logger,
+		quit:              make(chan bool),
+		toExitCh:          make(chan struct{}),
+		cfg:               cfg,
+		queue:             queue,
+		account:           account,
+		stateRootInHeader: chain.GetConfig().StateRootInHeader,
+	}
+}
+
+// Start implements the Service interface.
+func (bfs *Service) Start() {
+	if bfs.started.CompareAndSwap(false, true) {
+		bfs.log.Info("Starting Block Fetcher Service")
+		var err error
+		bfs.client, err = neofs.GetSDKClient(context.Background(), bfs.cfg.Addresses[0])
+		if err != nil {
+			bfs.log.Error("Failed to create SDK client", zap.Error(err))
+			close(bfs.quit)
+			return
+		}
+		defer bfs.client.Close()
+
+		go func() {
+			defer close(bfs.toExitCh)
+			switch bfs.cfg.Mode {
+			case ModeIndexSearch:
+				err = bfs.fetchData()
+			case ModeOidSearch:
+				err = bfs.fetchDataWithOid()
+			default:
+				bfs.log.Error("Invalid mode specified", zap.String("mode", bfs.cfg.Mode))
+				close(bfs.quit)
+				return
+			}
+
+			if err != nil {
+				bfs.log.Error("Fetch operation failed", zap.Error(err))
+				close(bfs.quit)
+				return
+			}
+		}()
+	}
+}
+
+// Shutdown implements the Service interface.
+func (bfs *Service) Shutdown() {
+	if bfs.started.CompareAndSwap(true, false) {
+		bfs.log.Info("Shutting down Block Fetcher Service")
+		close(bfs.quit)
+		<-bfs.toExitCh
+	}
+}
+
+// IsActive implements the Service interface.
+func (bfs *Service) IsActive() bool {
+	return bfs.started.Load()
+}
+
+func (bfs *Service) fetchData() error {
+	startIndex := bfs.chain.BlockHeight()
+	batchSize := uint32(bfs.cfg.BatchSize)
+
+	for {
+		select {
+		case <-bfs.quit:
+			bfs.log.Info("Stopping fetchData operation.")
+			return nil
+		default:
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(bfs.cfg.BlockAttribute, fmt.Sprintf("%d", startIndex), object.MatchNumGE)
+			filters.AddFilter(bfs.cfg.BlockAttribute, fmt.Sprintf("%d", startIndex+batchSize-1), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			blockOids, err := bfs.search(prm)
+			bfs.log.Info(fmt.Sprintf("Found %d blocks from index %d to %d", len(blockOids), startIndex, startIndex+batchSize-1))
+			if err != nil {
+				bfs.log.Error(fmt.Sprintf("Failed to find block with index %d", startIndex), zap.Error(err))
+				return err
+			}
+
+			if len(blockOids) == 0 {
+				bfs.log.Info(fmt.Sprintf("No block found with index %d, stopping.", startIndex))
+				return errors.New("no block found")
+			}
+			err = bfs.fetchAndProcessBlocks(blockOids)
+			if err != nil {
+				return err
+			}
+			startIndex += batchSize
+		}
+	}
+}
+
+func (bfs *Service) fetchAndProcessBlocks(blockOids []oid.ID) error {
+	var wg sync.WaitGroup
+
+	for _, oidBlock := range blockOids {
+		wg.Add(1)
+		go func(oidBlock oid.ID) {
```

Comment:
For `ModeOidSearch` Batch of block hashes is 2K, this code will cause memory consumption explosion. We need a batched downloader for `ModeOIDSearch`, the number of downloader threads should be configured via `bfw.cfg.BatchSize`.

I'd suggest to slightly modify `fetchDataWithOid` so that relatively small number of OIDs is requested from `fetchAndProcessBlocks`. E.g. make a cycle and split 2K OIDs into batches of `bfs.cfg.BatchSize` exactly like `fetchData` does it. These batches should be downloaded sequentially then.

`fetchData` doesn't have this problem.